### ISSUE TITLE
Epic: Differential testing of modularized service vs mono service

### ISSUE THREAD

**Issue by david-bakin-sl** [hashgraph/hedera-services#7619]
Soon will come the glorious day when we "flip the switch": we freeze and exit the mono service for the last time and start up the modular service.  At that point everything _had better work_: correctly, and forever. 

**Goal:** run one month of mainnet events through the modular code, then compare with the record files produced in mainnet. Ideally there are no differences. Any variations in the modular record need to be addressed by either filing a bug ticket or by documenting why the change is acceptable. 

**Record File Differential Test Setup:** see the instructions in [this](https://github.com/hashgraph/hedera-services/issues/10571) ticket for running modular record file generation, and for comparison of the result with a mono record file. 

We will build a differential testing facility that will run mono service and modularized service in parallel, going forward and on (recent) historical event streams, comparing the results for identity, and where they're different facilitating the developer intensive analysis and debugging of the modularized service.  The end result will be coverage of months worth of mainnet and testnet data (and maybe more); ensuring they're always equal for all real-world input.  The goal is to gain confidence in the new modularized service to the point where we can make a informed decision to deploy it.

This internal-only Notions page ["Differential Testing For Modularized Services vs Baseline "Mono" Service](https://www.notion.so/swirldslabs/Differential-Testing-for-Modularized-Services-vs-Baseline-Mono-Service-366106721f9a433eb8833408ecd436eb?pvs=4) - _though currently a work in progress_ - describes the effort generally.

This epic tracks the progress of the development of the diffferential testing facility through to preparations for its deployment.  (A subsequent epic, at the right time, will track progress in comparing the modular service to the mono service, with this tool.)

_**NOTE: lower-priority tasks have a down arrow by them**_ ⬇️

## The Primitive Building Block

The primitive building block is a command line tool (PCLI) + scripts that will take a 15-minute interval saved state to saved state, on the production mono service mainnet/testnet/etc. - fetch the bookend saved state files, all the input post-consensus event stream files and all the output record stream files for that interval, run the event stream files through the modularized service, capturing the resulting record stream files.

- [x] #10880
- [ ] #7716 @david-bakin-sl 
- [x] #11021 @tinker-michaelj 
- [x] #11006 
- [x] #11057 @mhess-swl 
- [ ] #7744 ⬇️
       - This is to support troubleshooting a record stream file difference 

## The Difference Engine

The difference engine uses the primitive building block to orchestrate two flows of tests:  All the intervals in the _present_ time, and as many intervals going back in time as we can run (before we get into incompatibilities in file formats or hedera semantics).  Running in the cloud (to reduce egress costs of all the data) it will run intervals and track results in a database of intervals.

- [x] #10084 – Migration prototype
- [x] #10571
- [x] ~~#10946~~
- [ ] #10879 @david-bakin-sl 
- [x] ~~Figure out why it takes 10-15 minutes to run 2 seconds' worth of events~~ – not needed now that we can run events with PCLI, which runs them _much_ faster.
- [x] #10950
- [x] #10949 @mhess-swl 
- [x] #11068
- [x] #10955
- [x] #10953
- [ ] TBD - Overall process document + supporting scripts and processes as necessary for entire "analytical" workflow

## The Analytical Engine

As the _difference engine_ produces results from an interval run by the modularized service, the _analytical engine_ will compare those results for equality.  If equal: great! If not equal it will go further and identify the difference (at the event level), bin it against other previously discovered differences _by similarity_ ("similarity" measures are not defined and will probably have to be discovered during use), and file issues.  When the issues are resolved it will trigger the difference engine to re-run the "fixed" intervals.  There'll be a dashboard or other reporting mechanism so progress can be tracked and developer resources assigned.

"Equality" is also something that may need tweaking as _expected_ differences in handling events, between the modularized service and the mono service, show up in the output stream.

There will also be scripts and/or tools to help the developer use the _primitive building block_ on an interval where a difference has been detected to drive the system to the exact event where the divergence takes place, to facilitate debugging.

**2024-01-10:** [This notions page](https://www.notion.so/swirldslabs/Signed-State-Store-Dumpers-57d4412fa4c84459a0ca699ce1446c98?pvs=4) "Signed State Store Dumpers" describes the work to be done to have the dumpers for both mono-and modular representation.  Issues will be updated shortly.

- [ ] Explore the possibility of integrating the dumpers into the node: the benefits, costs (things missing, is it as easy as we think, etc.) - report and decide what to do with states (leading to new issues)
- [x] All store dumper work now moved to #11207 - "Add/enhance/modify store dumpers for differential testing" - to make this issue more manageable

## Doing Differential Testing
* These tasks "run locally" are done _without_ comparing final state, only comparing record streams
* These tasks "run locally" are done until diminishing returns, then proceed to the next one
- [x] #11203
- [ ] #11356 
- [x] #11204
- [ ] #11326
- [ ] #11325
- [ ] Now run the full 30-days - comparing state at _end_
- [ ] Start testing state w/ state dumpers

## Miscellaneous _Potential_ Tasks
We're not sure if we need these tasks yet, but want to keep them in mind:
- [ ] Create a pre-migrated modular state (or states, depending on need) to use in performance testing
- [ ] Figure out how we want to cache state and stream files so we're not re-downloading them. Related to #7716 ⬇️
- [ ] #7742 ⬇️
- [ ] Write tests for the dumpers, or a single test for all the dumpers (e.g. on a known state)
